File: leetcode/java/removeNthFromEnd.java
/*
 * given a SLL, remove the Nth node from the end
 *
 * test data:
 * 1->2->3, 2
 * 1->2, 1
 * 1->2, 2
 * 2, 1
 * */
import java.io.*;
import java.util.*;

public class removeNthFromEnd{
    public class ListNode{
        int val;
        ListNode next;
        ListNode(int x){
            val = x;
            next = null;
        }
    }
    public ListNode removeNth2End(ListNode head, int n){
        if(head==null)    return null;
        if(n==1){
            ListNode prev = null, l=head;
            while(l.next != null){
                prev = l;
                l = l.next;
            }
            if(prev != null)    prev.next = null;
            return prev==null ? null : head;
        }
        
        ListNode p1 = head;
        for(int i=1;i<n && p1 != null; i++){
            p1 = p1.next;
        }
        if(p1==null)    return null;
        ListNode p2 = head, cprev = null;
        while(p1.next != null){
            cprev = p2;
            p2 = p2.next;
            p1 = p1.next;
        }
        ListNode h = head;
        if(cprev != null)    cprev.next = p2.next;
        else                 h = p2.next;
        p2.next = null;
        return h;
    }

    public static void main(String[] args){
        
    }
}

File: leetcode/src/removeNthFromEnd.cpp
/*
 * given a SLL, remove the Nth node from the end of list and return its head.
 *
 * test data:
 * 1->2->3->4->5, n=2, return head of SLL 1->2->3->5
 * 1, n=1,
 * 1->2, n=2
 * 1->2, n=1
 * */

#include "../include/preliminary.h"

class Solution{

public:
    /*
     * move p1 and p2 together where p2 of Nth back from p1
     * */
    ListNode* removeNthFromEnd(ListNode *head, int n){
        if(!head)    return NULL;

        ListNode *p1 = head; // p1 is the leading node
        for(int i = 1; i < n && p1 != NULL; i++){ // n > 1 so p1 and p2 cannot be same node
            p1 = p1->next;
        }

        if(!p1)    return NULL; // out of range

        ListNode *p2 = NULL, *prev = NULL;  // p2 is the Nth from p1, prev is pre of p2
        while(p1 != NULL){  // when p1 becomes NULL, p2 is the one to delete
            p1 = p1->next;
            prev = p2;
            p2 = ( !p2 ? head : p2->next);
        }

        if(prev != NULL){
            prev->next = p2->next;
        }else{ // prev is NULL, head is the one to delete
            head = p2->next;
        }

        p2->next = NULL;
        delete p2;

        return head;
    }
};

/* unit test is in ../cpp_unittest/removeNthFromEnd_unittest */
